src/rt/rwlock.rs

```diff
@@ -0,0 +1,269 @@
+                }
+
+                let obj = thread
+                    .operation
+                    .as_ref()
+                    .map(|operation| operation.object());
+
+                if obj == Some(self.obj) {
+                    thread.set_blocked();
+                }
+            }
+
+            state.lock = Some(RwLockType::Read(thread_set));
+
+            true
+        })
+    }
+
+    fn is_read_locked(&self) -> bool {
+        super::execution(
+            |execution| match self.get_state(&mut execution.objects).lock {
+                Some(RwLockType::Read(_)) => true,
+                _ => false,
+            },
+        )
+    }
+}
+
+impl RwLock {
+    /// Write RwLock functions
+    pub(crate) fn acquire_write_lock(&self) {
+        self.obj
+            .branch_acquire(self.is_write_locked() || self.is_read_locked());
+        assert!(
+            self.post_acquire_write_lock(),
+            "expected to be able to acquire write lock"
+        );
+    }
+
+    pub(crate) fn try_acquire_write_lock(&self) -> bool {
+        self.obj.branch_opaque();
+        self.post_acquire_write_lock()
+    }
+
+    pub(crate) fn release_write_lock(&self) {
+        super::execution(|execution| {
+            let state = self.get_state(&mut execution.objects);
+
+            state.lock = None;
+
+            state
+                .synchronize
+                .sync_store(&mut execution.threads, Release);
+
+            if state.seq_cst {
+                // Establish sequential consistency between the lock's operations.
+                execution.threads.seq_cst();
+            }
+
+            let thread_id = execution.threads.active_id();
+
+            // Block all other threads from attempting to acquire the RwLock
+            for (id, thread) in execution.threads.iter_mut() {
+                if id == thread_id {
+                    continue;
+                }
+
+                let obj = thread
+                    .operation
+                    .as_ref()
+                    .map(|operation| operation.object());
+
+                if obj == Some(self.obj) {
+                    thread.set_runnable();
+                }
+            }
+        });
+    }
+
+    fn post_acquire_write_lock(&self) -> bool {
+        super::execution(|execution| {
+            let state = self.get_state(&mut execution.objects);
+            let thread_id = execution.threads.active_id();
+
+            // Set the lock to the current thread
+            state.lock = match state.lock {
+                Some(RwLockType::Read(_)) => return false,
+                _ => Some(RwLockType::Write(thread_id)),
+            };
+
+            dbg!(state.synchronize.sync_load(&mut execution.threads, Acquire));
+
+            if state.seq_cst {
+                // Establish sequential consistency between locks
+                execution.threads.seq_cst();
+            }
+
+            // Block all other threads attempting to acquire rwlock
```

Comment:
Nit/TIOLI: seems like this code is repeated enough that it could maybe be factored out into its own function...not a blocker.